### PR TITLE
LoRA fixes for meta device initialization

### PR DESCRIPTION
**Pull request by carmocca**

### CODE REVIEW

**Comment by Andrei-Aksionov** [comment]
Just a question: is there a problem with registered buffer when the model was initialized on a meta device?

File: lit_gpt/lora.py [hunk 1]
```diff
@@ -589,15 +582,15 @@ def __init__(self, config: Config) -> None:
             config.n_embd,
             config.intermediate_size,
             bias=config.bias,
-            r=config.r,
+            r=(config.r if config.to_mlp else 0),
```
Comment (gkroiz) [inline lit_gpt/lora.py:585]: In `lora.py` (and maybe other files using LoRA), the default value that will set `config.to_mlp` is `False` (https://github.com/Lightning-AI/lit-gpt/blob/main/finetune/lora.py#L46). Do we want https://github.com/Lightning-AI/lit-gpt/blob/main/finetune/lora.py#L46 to be False by default?

Comment (carmocca) [reply lit_gpt/lora.py:585]: The original LoRA paper only applies it to Q, and V. Later papers suggest applying it to other layers so support was added by always converting the layer to LoRA but controling whether it's enabled via `r`.

So `to_mlp` should be kept as `False` to be consistent with the original LoRA paper.

This PR fixes that the MLP linear layers weren't getting converted as the rest of the linears.

**Comment by carmocca** [comment]
> Just a question: is there a problem with registered buffer when the model was initialized on a meta device?

Yes, you get a segfault error: 

```python
../aten/src/ATen/native/cuda/IndexKernel.cu:175: operator(): block: [3,0,0], thread: [27,0,0] Assertion `idx >= 0 && idx < self_dim_size && "index_copy_(): index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:175: operator(): block: [3,0,0], thread: [28,0,0] Assertion `idx >= 0 && idx < self_dim_size && "index_copy_(): index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:175: operator(): block: [3,0,0], thread: [29,0,0] Assertion `idx >= 0 && idx < self_dim_size && "index_copy_(): index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:175: operator(): block: [3,0,0], thread: [30,0,0] Assertion `idx >= 0 && idx < self_dim_size && "index_copy_(): index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:175: operator(): block: [3,0,0], thread: [31,0,0] Assertion `idx >= 0 && idx < self_dim_size && "index_copy_(): index out of bounds"` failed.
Traceback (most recent call last):
  File "/home/carlos/lit-parrot/thing.py", line 62, in <module>
    loss = model(x).sum()
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/lightning/fabric/wrappers.py", line 117, in forward
    output = self._forward_module(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/lit-parrot/lit_gpt/lora.py", line 517, in forward
    x, *_ = block(x, (cos, sin), max_seq_length)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 832, in forward
    output = self._fsdp_wrapped_module(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/lit-parrot/lit_gpt/model.py", line 168, in forward
    h, new_kv_cache = self.attn(n_1, rope, max_seq_length, mask, input_pos, kv_cache)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/lit-parrot/lit_gpt/model.py", line 205, in forward
    qkv = self.attn(x)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/carlos/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/carlos/lit-parrot/lit_gpt/lora.py", line 375, in forward
    result += self.zero_pad(after_B) * self.scaling  # (64, 64, 256) after zero_pad (64, 64, 384)
  File "/home/carlos/lit-parrot/lit_gpt/lora.py", line 319, in zero_pad
    result = result.index_copy(1, self.lora_ind.long(), x.reshape(-1, shape)) # (4096, 256)
RuntimeError: CUDA error: device-side assert triggered
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

[E ProcessGroupNCCL.cpp:848] [Rank 0] NCCL watchdog thread terminated with exception: CUDA error: device-side assert triggered
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Exception raised from c10_cuda_check_implementation at ../c10/cuda/CUDAException.cpp:44 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7f685248a4c7 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7f685244689b in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #2: c10::cuda::c10_cuda_check_implementation(int, char const*, char const*, int, bool) + 0x118 (0x7f6852537f78 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libc10_cuda.so)
frame #3: c10d::ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const + 0x80 (0x7f67f1bcce00 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #4: c10d::ProcessGroupNCCL::WorkNCCL::isCompleted() + 0x58 (0x7f67f1bd0cb8 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #5: c10d::ProcessGroupNCCL::workCleanupLoop() + 0x11b (0x7f67f1be6bcb in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #6: c10d::ProcessGroupNCCL::ncclCommWatchdog() + 0x78 (0x7f67f1be6f48 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #7: <unknown function> + 0xdc2b3 (0x7f6851cb22b3 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #8: <unknown function> + 0x94b43 (0x7f687bb08b43 in /lib/x86_64-linux-gnu/libc.so.6)
frame #9: <unknown function> + 0x126a00 (0x7f687bb9aa00 in /lib/x86_64-linux-gnu/libc.so.6)

terminate called after throwing an instance of 'std::runtime_error'
  what():  [Rank 0] NCCL watchdog thread terminated with exception: CUDA error: device-side assert triggered
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

Exception raised from c10_cuda_check_implementation at ../c10/cuda/CUDAException.cpp:44 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7f685248a4c7 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7f685244689b in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libc10.so)
frame #2: c10::cuda::c10_cuda_check_implementation(int, char const*, char const*, int, bool) + 0x118 (0x7f6852537f78 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libc10_cuda.so)
frame #3: c10d::ProcessGroupNCCL::WorkNCCL::finishedGPUExecutionInternal() const + 0x80 (0x7f67f1bcce00 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #4: c10d::ProcessGroupNCCL::WorkNCCL::isCompleted() + 0x58 (0x7f67f1bd0cb8 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #5: c10d::ProcessGroupNCCL::workCleanupLoop() + 0x11b (0x7f67f1be6bcb in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #6: c10d::ProcessGroupNCCL::ncclCommWatchdog() + 0x78 (0x7f67f1be6f48 in /home/carlos/venv/lib/python3.10/site-packages/torch/lib/libtorch_cuda.so)
frame #7: <unknown function> + 0xdc2b3 (0x7f6851cb22b3 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #8: <unknown function> + 0x94b43 (0x7f687bb08b43 in /lib/x86_64-linux-gnu/libc.so.6)
frame #9: <unknown function> + 0x126a00 (0x7f687bb9aa00 in /lib/x86_64-linux-gnu/libc.so.6)

Aborted (core dumped)
```

**Pull request merged**